stdlib/peval/ast.mc

```diff
@@ -0,0 +1,110 @@
+include "mexpr/keyword-maker.mc"
+include "mexpr/ast.mc"
+include "mexpr/ast-builder.mc"
+include "mexpr/type-check.mc"
+include "mexpr/type-annot.mc"
+include "mexpr/eval.mc"
+include "mexpr/info.mc"
+include "mexpr/eq.mc"
+
+include "peval/peval.mc"
+include "error.mc"
+include "list.mc"
+
+
+lang PEvalAst = KeywordMaker + MExpr + MExprEq + Eval + PrettyPrint
+                + MExprTypeCheck + LamEval + TypeAnnot + MExprPEval
+
+  syn Expr =
+  | TmPEval {e: Expr, info: Info}
+
+  -- States that the new terms are indeed mapping from keywords
+  sem isKeyword =
+  | TmPEval _ -> true
+
+  -- Defines the new mapping from keyword to new terms
+  sem matchKeywordString (info: Info) =
+  | "peval" -> Some (1, lam lst. TmPEval {e = get lst 0,
+                                          info = info})
+  sem tyTm = 
+  | TmPEval t -> tyTm t.e
+
+  sem infoTm =
+  | TmPEval t -> t.info
+
+  sem withType (ty : Type) =
+  | TmPEval t -> TmPEval {t with e = withType ty t.e}
+
+  sem typeCheckExpr (env : TCEnv) = 
+  | TmPEval t -> 
+    let e = typeCheckExpr env t.e in
+    TmPEval {t with e = e}
+
+  sem typeAnnotExpr (env : TypeEnv) = 
+  | TmPEval t -> 
+    let e = typeAnnotExpr env t.e in
+    TmPEval {t with e = e}
+
+  sem smapAccumL_Expr_Expr f acc =
+  | TmPEval t ->
+    match f acc t.e with (acc, e) in
+    (acc, TmPEval {t with e = e})
+
+  -- Equality of the new terms
+  sem eqExprH (env : EqEnv) (free : EqEnv) (lhs : Expr) =
+  | TmPEval r ->
+      match lhs with TmPEval l then
+        eqExprH env free l.e r.e
+      else None ()
+
+  sem eval (ctx : EvalCtx) = 
+  | TmPEval e -> 
+  switch eval ctx e.e 
+    case clos & TmClos {ident=i, body=b, env=envi} then
+        match peval clos with res then
+            match res with TmLam t then
+                eval ctx res -- TmClos ...
+            else res
+        else never
```

Comment:
There is a syntactic sugar `match e1 with p in e2` = `match e1 with p then e2 else never` if you prefer this